src/content/docs/guides/Beginners/Your First Shader/0_intro.mdx

```diff
@@ -0,0 +1,57 @@
+
+Lighting in Minecraft is a bit different from how it is done in other games. Minecraft needs to support an arbitrary number of light sources, with the features of old OpenGL versions, and have decent performance on slow hardware like Intel iGPUs. There also needs to be occlusion detection for the lights, that is, a light behind a wall cannot light up what is in front of the the wall. Doing this the “normal” way would require storing all lights in a texture and having a texture atlas of shadow maps for each light. This doesn't support area lighting, so lighting from blocks like glowstone up close will look bad, and this would be insanely costly. Imagine how slow rendering the nether would be, since each lava block in the nether needs to be processed. Minecraft needs a different approach from this.
+
+Some of you who play Minecraft will know that each block has a lighting level, which comes from both torches and how exposed a block is to the sky. Minecraft reuses this information for lighting the blocks. Each vertex has a `vec2` attribute known as the “lightmap coordinates”. The x value represents lighting from blocks like torches and glowstone, while the y value represents how much the vertex is exposed to the sky. These values in older versions of Minecraft are from 0 to 15, but in newer versions it can be up to the 200s.
+
+The lightmap alone is not enough to light the block. It somehow has to be converted to a lighting color which then has to be multiplied by the block color to obtain the final color that gets displayed on your screen. Minecraft by default uses the light map coordinates (after doing math to move them to the [0, 1] range) as texture coordinates to look up a lighting color value from a lightmap texture in the fragment shader. The lighting color value gets multiplied by the block color and then displayed on your screen. See the [Optifine documentation on this](https://github.com/sp614x/optifine/blob/15ef31064323d7e1c5959ab8f9e8a260f0750124/OptiFineDoc/doc/shaders.txt#L263) for more details. We won't be using the light map coordinates to look up from the lightmap texture. We will be doing our own math to calculate the lighting value. This is what most other shaders do anyway (and when I mean most I mean 99.9%, if it doesn't, then it's probably a really bad or super old no-name shader).
+
+## How Shaders Work
+
+To understand how shaders work, let's understand how the shader pipeline works. The majority of shader programs used are 'fullscreen passes'. These are passes which run for every pixel on the screen. These are the simplest form of a shader program, and are very useful for post processing effects, or anything which doesn't require information about what's not on screen.
```

Comment:
> The majority of shader programs used are 'fullscreen passes'
I wouldn't necessarily say this is true (forward rendered shaders are a thing and are probably more common than you'd think), and regardless the line isn't really useful, the rest of the paragraph explains them pretty well (although maybe it would be good to mention deferred rendering)

Reply:
Removed that claim. Deferred rendering is mentioned in the next page.